examples/sampling/pyg/node_classification.py

```diff
@@ -0,0 +1,210 @@
+This approach not only underscores the adaptability and scalability of modern graph neural network frameworks but also opens new avenues for 
+handling and analyzing large graph datasets in various real-world applications.
+
+Key Features:
+- Implements the GraphSAGE model, a scalable GNN, for node classification on large graphs.
+- Utilizes GraphBolt, an efficient framework for large-scale graph data processing.
+- Integrates with PyTorch Geometric for building and training the GraphSAGE model.
+- The script is well-documented, providing clear explanations at each step.
+
+This flowchart describes the main functional sequence of the provided example.
+main: 
+
+main
+│
+├───> Load and preprocess dataset (GraphBolt)
+│     │
+│     └───> Utilize GraphBolt's BuiltinDataset for dataset handling
+│
+├───> Instantiate the SAGE model (PyTorch Geometric)
+│     │
+│     └───> Define the GraphSAGE model architecture
+│
+├───> Train the model
+│     │
+│     ├───> Mini-Batch Processing with GraphBolt
+│     │     │
+│     │     └───> Efficient handling of mini-batches using GraphBolt's utilities
+│     │
+│     └───> Training Loop
+│           │
+│           ├───> Forward and backward passes
+│           │
+│           └───> Parameters optimization
+│
+└───> Evaluate the model
+      │
+      └───> Performance assessment on validation and test datasets
+            │
+            └───> Accuracy and other relevant metrics calculation
+
+
+"""
+
+import dgl.graphbolt as gb
+import torch
+import torch.nn.functional as F
+from torch_geometric.nn import SAGEConv
+
+
+class GraphSAGE(torch.nn.Module):
+    #####################################################################
+    # (HIGHLIGHT) Define the GraphSAGE model architecture.
+    #
+    # - This class inherits from `torch.nn.Module`.
+    # - Two convolutional layers are created using the SAGEConv class from PyG.
+    # - 'in_size', 'hidden_size', 'out_size' are the sizes of
+    #   the input, hidden, and output features, respectively.
+    # - The forward method defines the computation performed at every call.
+    #####################################################################
+    def __init__(self, in_size, hidden_size, out_size):
+        super(GraphSAGE, self).__init__()
+        self.layers = torch.nn.ModuleList()
+        self.layers.append(SAGEConv(in_size, hidden_size))
+        self.layers.append(SAGEConv(hidden_size, hidden_size))
+        self.layers.append(SAGEConv(hidden_size, out_size))
+
+    def forward(self, blocks, x, device):
+        h = x
+        for i, (layer, block) in enumerate(zip(self.layers, blocks)):
+            edge_index = block.edges()
+            edge_index = torch.stack([edge_index[0], edge_index[1]], dim=0)
+            h = layer(h, edge_index)
+            if i != len(blocks) - 1:
+                h = F.relu(h)
+            h = h[: block.number_of_dst_nodes()]
```

Comment:
this aims to extract dst nodes' output feature? then should it be `h[block.dstdata[dgl.NID]]`? As this is in model forward and it's for PyG, I think we'd better convert DGLBlocks into approriate data structures that could run with PyG model. just avoid DGL code within PyG model.

Reply:
if we do h[block.dstdata[dgl.NID]], it would cause error of KeyError: '_ID'. I think h[block.dstdata[dgl.NID]] is for DGL model, but for PyG, block.number_of_dst_nodes() is enough to extract output features from Graphbolt.

Reply:
if we do h[block.dstdata[dgl.NID]], it would cause error of KeyError: '_ID'. I think h[block.dstdata[dgl.NID]] is for DGL model, but for PyG, block.number_of_dst_nodes() is enough to extract output features.

Reply:
Ah, `h[block.dstdata[dgl.NID]]` is incorrect. why do we need to call `h[: block.number_of_dst_nodes()]`? num_nodes mismatch with next layer's input nodes? what I am concerning is whether the sliced `h` match with the input nodes in next layer.

Reply:
pls add a comment here to explain why we need to slice it as the first xxx nodes are seed nodes and those feature data should be passed to next layer.

Reply:
https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.SAGEConv.html
They explain here that there are two modes, normal or bipartite. In bipartite mode, we don't need to slice but need to pass input features separately for source and destination vertices, as far as I understand.

Reply:
@Rhett-Ying @hutiechuan our case should be bipartite as we are working with blocks, I am guessing.

Reply:
The thing is that if I don't slice, it would cause the error :

File "/home/ubuntu/dgl/examples/sampling/pyg/node_classification.py", line 268, in <module>
    main()
  File "/home/ubuntu/dgl/examples/sampling/pyg/node_classification.py", line 254, in main
    train_loss, train_accuracy = train(
  File "/home/ubuntu/dgl/examples/sampling/pyg/node_classification.py", line 164, in train
    loss = criterion(out, labels)
  File "/opt/conda/envs/dgl-dev-gpu-121/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/dgl-dev-gpu-121/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/dgl-dev-gpu-121/lib/python3.8/site-packages/torch/nn/modules/loss.py", line 1179, in forward
    return F.cross_entropy(input, target, weight=self.weight,
  File "/opt/conda/envs/dgl-dev-gpu-121/lib/python3.8/site-packages/torch/nn/functional.py", line 3053, in cross_entropy
    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
ValueError: Expected input batch_size (28529) to match target batch_size (1024).

My guess is that heterogeneous Graphs in PyG when dealing with heterogeneous graphs, the framework can automatically handle different node types and the relationships between them. However, this doesn't necessarily mean that slicing is not needed.

Reply:
This example is not heterogeneous?